### PR TITLE
kv request: configurable KV Timeout 

### LINKED ISSUE #45380: Configurable KV Timeout Tracking Issue
## Background

- As described in the [proposal](https://github.com/pingcap/tidb/blob/7105505a78fc886c33258caa5813baf197b15247/docs/design/2023-06-30-configurable-kv-timeout.md), to mitigate the impact of latency jitter encountered by the tikv nodes.
- The related feature request [issue](https://github.com/pingcap/tidb/issues/44771).

### TiDB

- [ ] Add new session variable `tidb_kv_read_timeout`. #45601
- [ ] Support hint processing for `tidb_kv_read_timeout`. When parsing and processing user hints in the `ExtractTableHintsFromStmtNode` and `handleStmtHints` functions, convert the input hint value into the corresponding field of `StmtContext` so it can be used later. #45601
- [ ] Add the `KVReadTimeout` field in the `StatementContext`, it could be set from the session variable or the hint value. #45601
- [ ] Support timeout processing logic in the `pointGetExecutor` and `batchPointGetExecutor`, setting the timeout value to the underlying `KVSnapshot` which is used by the executors.#45601
- [ ] Support timeout processing logic in the `copTask`, setting the timeout value to the `Context.max_execution_duration_ms` field when building coprocessor kv-reuqests. #45601

### KV-Client

- [ ] Add set interface `SetKVReadTimeout` and get interface `GetKVReadTimeout` for `KVSnapshot`. So the executors in TiDB could use them to get or set timeout values. Setting the `tidb_kv_read_timeout` in the `Context.max_execution_duration_ms` when `get` an `batchGet` are used. https://github.com/tikv/client-go/pull/919
- [ ] Support specific timeout retry strategy in the KV-Client for the `tidb_kv_read_timeout`. When a timeout happens, retry the next available peer for the read requests like stale read and non-leader-only snapshot read requests. The strategy is:
  - Create a specific error type for this kind of timeout, for example, `ConfigurableReadTimeoutErr`. It should not be mixed with the existing `RPCError`. https://github.com/tikv/client-go/pull/919
  - Try the next peer if the first request times out immediately. https://github.com/tikv/client-go/pull/919
  - If all the responses from all the peers are timeout errors, return timeout error to the upper layer, Backoff with the timeout error, and retry the requests with the original default `ReadShortTimeout` and `ReadMediumTimeout` values. Continue to process like before. https://github.com/tikv/client-go/pull/919
  - The timeout retry details should be recorded and handled properly in the request tracker, and so are the metrics.
  - Add related retry and statement management logic for `KVSnapshot` and `ClientHelper`.

### TiKV

- [ ] Support timeout check during the request handling in TiKV. When there's new point get and batch point get 
requests are received, the `kv_read_timeout` value should be read from the `GetReuqest` and `BatchGetRequest`
requests and passed to the `pointGetter`. By now there's no canceling mechanism when the task is scheduled to the
read pool in TiKV, a simpler way is to check the deadline duration before the next processing and try to return
directly if the deadline is exceeded, but once the reading task is spawned to the read pool it could not be 
canceled anymore. https://github.com/tikv/tikv/pull/15307
- [x] <del>Support Timeout Configuration For Coprocessor Tasks. Use the `kv_read_timeout` value passed in to calculate the `deadline` result in `parse_and_handle_unary_request` instead of using the default one.</del>
  - Coprocessor task already support kv_read_timeout.

### PR DESCRIPTION
**Pull request by hihihuhu**
### What problem does this PR solve?

Issue Number: close #45380

Problem Summary:

### What is changed and how it works?

include:
 

- [x] Add new session variable tidb_kv_read_timeout.
- [x]  Support hint processing for tidb_kv_read_timeout. When parsing and processing user hints in the ExtractTableHintsFromStmtNode and handleStmtHints functions, convert the input hint value into the corresponding field of StmtContext so it can be used later.
- [x]  Add the KVReadTimeout field in the StatementContext, it could be set from the session variable or the hint value.
- [x]  Support timeout processing logic in the copTask, setting the timeout value to the Context.max_execution_duration_ms field when building coprocessor kv-reuqests.

not yet finished:
- [ ]  Support timeout processing logic in the pointGetExecutor and batchPointGetExecutor, setting the timeout value to the underlying KVSnapshot which is used by the executors.
waiting for client-go implementation, adding a TODO for now

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

cc @Tema

### CODE REVIEW

**Comment by hawkingrei** [comment]
@hihihuhu Can you change the target base branch to master?

**Comment by cfzjywxk** [comment]
> @hihihuhu Can you change the target base branch to master?

@hawkingrei Currently, we are actively developing in the development branch, and once the features are fully refined, they will be merged into the master branch.

**Comment by hawkingrei** [comment]
@hihihuhu Hi, I heard that you are updating dependencies. Now, the dependencies for TiDB are uploaded to a mirror, making the download faster and more stable. However, this operation requires permission. So, if you want to upgrade the dependencies, simply run make bazel_prepare. If you encounter a 404 error, you don't need to worry about it. Someone will help you with the upload. If it's urgent, you can mention me (@mention) for assistance.

**Comment by hawkingrei** [comment]
@hihihuhu Feel free to consult me for any development-related questions.

File: executor/adapter.go [hunk 1]
```diff
@@ -1779,7 +1779,7 @@ func getEncodedPlan(stmtCtx *stmtctx.StatementContext, genHint bool) (encodedPla
 			// some hints like 'memory_quota' cannot be extracted from the PhysicalPlan directly,
 			// so we have to iterate all hints from the customer and keep some other necessary hints.
 			switch tableHint.HintName.L {
-			case "memory_quota", "use_toja", "no_index_merge", "max_execution_time",
+			case "memory_quota", "use_toja", "no_index_merge", "max_execution_time", "tidb_kv_read_timeout",
```
Comment (MyonKeminta) [inline executor/adapter.go:1782]: Could we consider define a constant in `plannercore` scope, like the following several hints?

File: session/sessiontest/session_test.go [hunk 2]
```diff
@@ -1026,6 +1026,42 @@ func TestMaxExecutionTime(t *testing.T) {
...
+
+	tk.MustQuery("select @@TIDB_KV_READ_TIMEOUT;").Check(testkit.Rows("0"))
+	tk.MustQuery("select @@global.TIDB_KV_READ_TIMEOUT;").Check(testkit.Rows("0"))
+	tk.MustQuery("select /*+ TIDB_KV_READ_TIMEOUT(1000) */ * FROM TidbKvReadTimeout;")
+
+	tk.MustExec("set @@global.TIDB_KV_READ_TIMEOUT = 300;")
+	tk.MustQuery("select * FROM TidbKvReadTimeout;")
+
+	tk.MustExec("set @@TIDB_KV_READ_TIMEOUT = 150;")
+	tk.MustQuery("select * FROM TidbKvReadTimeout;")
```
Comment (MyonKeminta) [inline session/sessiontest/session_test.go:1054]: As the related functionality seems to be not yet implemented in client-go and TiKV, is these test code for ensuring that setting the variables or hint won't cause the statement crash?
It's good enough for the time being though, perhaps you can also try to check that the timeout are successfully passed to the `store/driver` layer.

**Review by MyonKeminta — COMMENTED** [review]
Nice work!

**Comment by hihihuhu** [comment]
@MyonKeminta thanks for the review! i have updated the dependency, could you please help to publish the client-go change so that the build job could pick it up from the cache server? thanks!

**Comment by hawkingrei** [comment]
> @MyonKeminta thanks for the review! i have updated the dependency, could you please help to publish the client-go change so that the build job could pick it up from the cache server? thanks!

I have uploaded it and restarted ci.

**Comment by hawkingrei** [comment]
@hihihuhu 

1、Please run the ```make bazel_prepare``` to upgrade the bazel config.
2、fix the linter.

**Comment by MyonKeminta** [comment]
> @hihihuhu
> 
> 1、Please run the `make bazel_prepare` to upgrade the bazel config. 2、fix the linter.

By executing `git apply -` and paste following content, `make bazel_prepare` should be able to be skipped. It can be found in the log of `idc-jenkins-ci-tidb/unit-test`.

```
diff --git a/distsql/BUILD.bazel b/distsql/BUILD.bazel
index 0a2b89cd5..3a4044139 100644
--- a/distsql/BUILD.bazel
+++ b/distsql/BUILD.bazel
@@ -65,7 +65,7 @@ go_test(
     embed = [":distsql"],
     flaky = True,
     race = "on",
-    shard_count = 23,
+    shard_count = 24,
     deps = [
         "//domain/resourcegroup",
         "//kv",
```

**Comment by crazycs520** [comment]
> 

I will run `make bazel_prepare`, @hihihuhu you don't need to care about this.

File: planner/optimize.go [hunk 3]
```diff
@@ -818,6 +821,16 @@ func handleStmtHints(hints []*ast.TableOptimizerHint) (stmtHints stmtctx.StmtHin
...
 	}
+	// Handle TIDB_KV_READ_TIMEOUT
+	if tidbKvReadTimeoutCnt != 0 {
+		tidbKvReadTimeout := hints[hintOffs["tidb_kv_read_timeout"]]
+		if tidbKvReadTimeoutCnt > 1 {
+			warn := errors.Errorf("TIDB_KV_READ_TIMEOUT() is defined more than once, only the last definition takes effect: TIDB_KV_READ_TIMEOUT(%v)", tidbKvReadTimeout.HintData.(uint64))
+			warns = append(warns, warn)
+		}
+		stmtHints.HasTidbKvReadTimeout = true
+		stmtHints.TidbKvReadTimeout = tidbKvReadTimeout.HintData.(uint64)
```
[the remark this answers is not in the document]
Comment (you06) [reply planner/optimize.go:832]: Details in this issue #45812

**Review by crazycs520 — APPROVED** [review]

**Comment by easonn7** [comment]
/approve

discussed with @cfzjywxk , LGTM

**Pull request merged**